### PR TITLE
Release 0.5.0 and some doc clarifications

### PR DESCRIPTION
**Pull request by gracinet**
Tried to make some things more explicit in the documentation of properties, and then proceed to the 0.5.0 release.

I'll merge this myself, so that I can upload to `crates.io` and put a tag right away.

### CODE REVIEW

**Comment by gracinet** [comment]
@indygreg here we go

@markbt hope what I'm writing about properties suits you.

**Comment by gracinet** [comment]
> For the property documentation (and for the class documentation overall), the main improvement I'd make is to call out how things are mapped to the slots in `PyTypeObject`. e.g. do Rust properties get mapped to `tp_getattro`, `tp_descr_get`, `tp_getset`, etc? But this is a very nitpicky doc comment and shouldn't block the release :)

I've considered doing that, partly because it's after all easier to just write "backed by `tp_getset`" (that's what it is btw).

Then I've told myself that the whole `py_class` documentation wasn't relying on of those C API names (slots, types etc.) to explain things. The only exceptions are the GC support methods, but they are explained entirely and could have been named arbitrarily, so I decided to do the same.

I take that as a deliberate sign that you don't have to be a C extension author to write an extension with the `cpython` crate.

That being said, a table summarizing for C extension authors how to translate the C API concepts that are familiar to them in the rust-cpython world would be a nice addition – that we could do later.

**Comment by gracinet** [comment]
I see that we have build failures on Rust nightly that are at first glance unrelated. I'll take care of these tomorrow (CEST), and then proceed to the release.

**Comment by gracinet** [comment]
I find it probable that the test failure is due to a bug in `rustc -Z ast-json`.
I've reproduced it also in #216, which does not change the code at all.

@markbt @dgrunwald a piece of advice would be useful to confirm that the test failure on nightly should not block the release.

It seems that the current `rustc -Z ast-json` has some invalid JSON output (it's actually a missing `id` value). This is part of `check_symbols.py`, which can run on nightly only (because `ast-json` isn't stable), yet seems to perform some important checks.

Here's an extract around the first occurrence of that missing `id` in the JSON. It's right after `"args":"Empty"`

```
"segments":[{"ident":{"name":"macro_use","span":{"lo":0,"hi":0}},"id":10,"args":null}]},"args":"Empty"}]},"id":,"style":"Outer","span":{"lo":0,"hi":0}}],"id":11,"span":{"lo":0,"hi":0},"vis":{"node":"Inherited","span":{"lo":0,"hi":0}},
```
If I perform a stupid string replacement with `"id": "MISSING"`, so that it's valid JSON again, then the test passes.

So, it looks to me actually as a bug in `-Z ast-json`: as long as the source is syntactically correct, which it is, we shouldn't get invalid JSON. Whereas this could be shadowing a real problem is what I'm not 100% sure - but our test passes if we force the JSON back into validity.

**Comment by indygreg** [comment]
The CI failure is a non-determinism issue in CI involving Rust Nightly. As we can see from the green checkboxes at https://github.com/dgrunwald/rust-cpython/commits/master, the last commit to the repo passed all CI 20 days ago. Since this PR isn't materially changing the code, I think it is safe to assume the CI checks would continue to pass if Nightly were stable. And I don't think the state `master` should be of concern.

File: src/py_class/py_class.rs [hunk 1]
```diff
@@ -171,16 +171,27 @@ Declares a static method callable from Python.
...
 * Unlike Python, the setter method name must be different from the property name.
   The setter method name is used to call the setter from Rust.
-* If value is `None` then the property is being deleted.
+* A `None` value represents a deletion call (typically Python `del` or `delattr()`).
+  As with Python properties, what should happen is entirely up to the implementation.
 * The value type can be any type that implements `FromPyObject`, or a reference or
   optional reference to any type that implements `RefFromPyObject`.  In the latter
   case, the type of the value is `Option<Option<&impl RefFromPyObject>>`, where
-  `None` means the property is being deleted, `Some(None)` means the property is
+  `None` means a deletion call, `Some(None)` means the property is
```
[the remark this answers is not in the document]
Comment (gracinet) [reply src/py_class/py_class.rs:194]: The thing is "is being deleted" is precisely what created confusion for me before I looked closer, but it's true that I hadn't remarked yet that this whole bullet list was about the setter only, and the title helps with that.

Would you agree with this: at first occurrence (L189),  with the added example. "A `None` value represents that the property is being deleted, for instance with Python `del`" ?

File: src/py_class/py_class.rs [hunk 2]
```diff
@@ -171,16 +171,27 @@ Declares a static method callable from Python.
...
-* The setter is optional.  If omitted, the attribute will be read-only.
+These "Rust properties" are not
+[Python properties](https://docs.python.org/3/library/functions.html#property),
+although they share some similarities with them: from the Python side, they behave like
+the attributes of some standard CPython library types, such as `datetime.date`
+(read-only) or `decimal.Context` (writable).
+
+### Setter details
+
+* The setter is optional.  If omitted, the attribute will be read-only, any setting or deleting
```
[the remark this answers is not in the document]
Comment (gracinet) [reply src/py_class/py_class.rs:185]: Ah thanks, always wondering about these.

File: src/py_class/py_class.rs [hunk 3]
```diff
@@ -171,16 +171,27 @@ Declares a static method callable from Python.
...
-Declares a property (attribute with code for getting and optionally setting) accessible from Python.
+Declares a Python data attribute backed by a Rust methods to
+get its value and, optionally, to set or delete it.
 
-* The setter is optional.  If omitted, the attribute will be read-only.
+These "Rust properties" are not
+[Python properties](https://docs.python.org/3/library/functions.html#property),
+although they share some similarities with them: from the Python side, they behave like
+the attributes of some standard CPython library types, such as `datetime.date`
+(read-only) or `decimal.Context` (writable).
```
[the remark this answers is not in the document]
Comment (gracinet) [reply src/py_class/py_class.rs:181]: Yes maybe I was overdoing it, I've kept the paragraph above, and I think it's enough for the layman.

**Comment by gracinet** [comment]
@markbt, does the version I just force pushed suit you or do you prefer not any doc change?

**Comment by markbt** [comment]
Those changes look good to me. Thanks! 
null

**Comment by gracinet** [comment]
Thanks, I'm about to proceed.

**Comment by gracinet** [comment]
Uploads to crates.io and tag also done.

**Pull request merged**